src/main/java/org/broadinstitute/hellbender/utils/tsv/SimpleCSVWriterWrapperWithHeader.java

```diff
@@ -0,0 +1,197 @@
+ *
+ * The expected use case for this class is that first {@link #setHeaderLine} is called with a list of the column names
+ * which will be used to determine the number of columns per line as well as how the header is indexed. Then in order to
+ * construct a new line call {@link #getNewLineBuilder} to get a line builder for each line, which then has convienent
+ * methods for individually assigning column values based on the header line etc. Once a line is finished being mutated
+ * one simply needs to call write() on the line to validate and finalize the line.
+ *
+ * Why didn't I use a tableWriter here? Who really holds the patent on the wheel anyway? Certainly not TableWriter.
+ */
+class SimpleCSVWriterWrapperWithHeader implements Closeable {
+    private int expectedColumns;
+    private Map<String, Integer> headerMap = null;
+    private CSVWriter outputWriter;
+
+    // The current incomplete line in the writer.
+    private SimpleCSVWriterLineBuilder currentLine = null;
+
+    /**
+     * Creates a new table writer given the file and column names.
+     *
+     * @param path         the destination path.
+     * @param separator    separator to use for the TSV file
+     * @throws IOException              if one was raised when opening the the destination file for writing.
+     */
+    public SimpleCSVWriterWrapperWithHeader(final Path path, final char separator) throws IOException {
+        this( new OutputStreamWriter(
+                Files.newOutputStream(Utils.nonNull(path, "The path cannot be null."))),
+        separator);
+    }
+
+    /**
+     * Creates a new table writer given the file and column names.
+     *
+     * @param writer       the destination writer.
+     * @param separator    separator to use for the TSV file
+     * @throws IOException              if one was raised when opening the the destination file for writing.
+     */
+    public SimpleCSVWriterWrapperWithHeader(final Writer writer, final char separator) {
+        outputWriter = new CSVWriter(writer, separator);
+    }
+
+    /**
+     * Provides a header line to the CSV output file. Note that this will throw an exception if all header lines
+     * are not unique as it attempts to create an index for the provided header lines for convenience when building
+     * rows of the CSV.
+     *
+     * @param columns Ordered list of header lines to be built into the CSV
+     */
+    public void setHeaderLine(List<String> columns) {
+        if (headerMap != null) {
+            throw new GATKException("Should not be adding multiple header lines to a file");
+        }
+        outputWriter.writeNext(columns.toArray(new String[0]), false);
+        expectedColumns = columns.size();
+
+        // Create the mapping between header and column
+        headerMap = new HashMap<>();
+        for (int i = 0; i < columns.size(); i++) {
+            Utils.nonNull(columns.get(i));
```

Comment:
Throw a more informative message here like "Column heading i was null" instead of using the generic `Utils.nonNull()`